File: frontend/src/mocks/fixtures/modelRegistry.ts
import type {
  ModelRegistryItem,
  ModelRegistryResponse,
} from '../../contracts/modelRegistry'

const sharedFacts: Pick<
  ModelRegistryItem,
  'dataset_reference' | 'window_size' | 'features' | 'score_semantics' | 'report_source'
> = {
  dataset_reference: 'b02f3872_ruang_produksi_v3_march07',
  window_size: 10,
  features: ['suhu', 'rh'],
  score_semantics: 'window_mean_squared_reconstruction_error',
  report_source: 'reported_model_registry',
}

export const modelRegistryResponse = {
  items: [
    {
      ...sharedFacts,
      id: 'conv1d_step5',
      family: 'conv1d',
      display_name: 'Conv1D Autoencoder',
      architecture: { latent_channels: 16 },
      param_count: 7_474,
      best_val_mse: 2.1572509291888413e-05,
      best_epoch: 4,
      model_sha256: '85c901e8fed463207a44151adc14772d3660384ae88daf9fcc53431e6acc39c9',
      summary: 'Conv1D reconstruction model dengan metrik validasi yang dilaporkan dari training.',
    },
    {
      ...sharedFacts,
      id: 'gru_step5',
      family: 'gru',
      display_name: 'GRU Autoencoder',
      architecture: { hidden_size: 32, latent_size: 8, layers: 2, dropout: 0.1 },
      param_count: 20_490,
      best_val_mse: 6.004524724196261e-05,
      best_epoch: 13,
      model_sha256: '0506d1da27d92a259e62c32ce43db7fd19dfa8ad679c08c6d67bf727653a2caa',
      summary: 'GRU reconstruction model dengan metrik validasi yang dilaporkan dari training.',
    },
    {
      ...sharedFacts,
      id: 'lstm_step5',
      family: 'lstm',
      display_name: 'LSTM Autoencoder',
      architecture: { hidden_size: 32, latent_size: 8, layers: 2, dropout: 0.1 },
      param_count: 27_210,
      best_val_mse: 5.146170129209432e-05,
      best_epoch: 24,
      model_sha256: '0dde621c1fe4117fd57602a94c30bd764e900108ceea3675fba6295e9500cccb',
      summary: 'LSTM reconstruction model dengan metrik validasi yang dilaporkan dari training.',
    },
    {
      ...sharedFacts,
      id: 'rnn_step5',
      family: 'rnn',
      display_name: 'RNN Autoencoder',
      architecture: { hidden_size: 32, latent_size: 8, layers: 2, dropout: 0.1 },
      param_count: 7_050,
      best_val_mse: 3.3277092602658214e-05,
      best_epoch: 15,
      model_sha256: 'c801a284c95c16ce9031a24f774d941c314bc0758e7b20d593af64fb630f0ebd',
      summary: 'RNN reconstruction model dengan metrik validasi yang dilaporkan dari training.',
    },
    {
      ...sharedFacts,
      id: 'transformer_step5',
      family: 'transformer',
      display_name: 'Transformer Autoencoder',
      architecture: {
        d_model: 32,
        n_heads: 4,
        encoder_layers: 2,
        decoder_layers: 2,
        ff_dim: 64,
        dropout: 0.1,
      },
      param_count: 43_362,
      best_val_mse: 3.5587262735700976e-05,
      best_epoch: 17,
      model_sha256: '364b0c73be1054b05a33924615d53ee1ebcb12af4bbb7d4efc0c1a144af3e015',
      summary: 'Transformer reconstruction model dengan metrik validasi yang dilaporkan dari training.',
    },
  ],
} satisfies ModelRegistryResponse

File: backend/anomaly_backend/model_registry.py
from __future__ import annotations

import hashlib
import json
from importlib import resources
from typing import cast

from pydantic import ValidationError

from anomaly_backend.contracts import ModelRegistryItem, ModelRegistryResponse

FIXTURE_PATH = "fixtures/model_registry/reported_models.json"
FIXTURE_SHA256 = "b0dcd9a5b44a8daf3d6ecc92d5b5c80c077c0ebcb0bc6fb412041b5e9756cdbe"

_RECURRENT_ARCHITECTURE: dict[str, int | float] = {
    "hidden_size": 32,
    "latent_size": 8,
    "layers": 2,
    "dropout": 0.1,
}

_EXPECTED_MODELS: dict[str, tuple[str, dict[str, int | float]]] = {
    "conv1d_step5": ("conv1d", {"latent_channels": 16}),
    "gru_step5": ("gru", dict(_RECURRENT_ARCHITECTURE)),
    "lstm_step5": ("lstm", dict(_RECURRENT_ARCHITECTURE)),
    "rnn_step5": ("rnn", dict(_RECURRENT_ARCHITECTURE)),
    "transformer_step5": (
        "transformer",
        {
            "d_model": 32,
            "n_heads": 4,
            "encoder_layers": 2,
            "decoder_layers": 2,
            "ff_dim": 64,
            "dropout": 0.1,
        },
    ),
}


class ModelRegistryIntegrityError(ValueError):
    pass


def normalize_reported_models(payload: object) -> list[ModelRegistryItem]:
    try:
        response = ModelRegistryResponse.model_validate(payload, strict=True)
    except ValidationError as error:
        raise ModelRegistryIntegrityError(
            "reported model registry has an invalid shape"
        ) from error

    if [item.id for item in response.items] != list(_EXPECTED_MODELS):
        raise ModelRegistryIntegrityError(
            "reported model registry must contain the five expected models"
        )
    for item in response.items:
        family, architecture = _EXPECTED_MODELS[item.id]
        if item.family != family or item.architecture != architecture:
            raise ModelRegistryIntegrityError(
                f"reported model registry architecture mismatch for {item.id}"
            )
    return response.items


def load_reported_models() -> list[ModelRegistryItem]:
    payload = resources.files("anomaly_backend").joinpath(FIXTURE_PATH).read_bytes()
    if hashlib.sha256(payload).hexdigest() != FIXTURE_SHA256:
        raise ModelRegistryIntegrityError(
            f"{FIXTURE_PATH} SHA-256 mismatch: expected {FIXTURE_SHA256}"
        )
    try:
        parsed = cast(object, json.loads(payload))
    except json.JSONDecodeError as error:
        raise ModelRegistryIntegrityError(
            f"{FIXTURE_PATH} must contain valid JSON"
        ) from error
    return normalize_reported_models(parsed)
